libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -0,0 +1,310 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_table_slice.hpp>
+#include <tenzir/collect.hpp>
+#include <tenzir/concept/parseable/tenzir/pipeline.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/tql/parser.hpp>
+
+namespace tenzir::plugins::deduplicate {
+
+namespace {
+
+using std::chrono::steady_clock;
+
+struct configuration {
+  friend auto inspect(auto& f, configuration& x) -> bool {
+    return f.object(x).fields(f.field("fields", x.fields),
+                              f.field("limit", x.limit),
+                              f.field("distance", x.distance),
+                              f.field("timeout", x.timeout));
+  }
+
+  std::vector<std::string> fields{};
+  int64_t limit{};
+  int64_t distance{};
+  steady_clock::duration timeout{};
+};
+
+class deduplicate_operator final : public crtp_operator<deduplicate_operator> {
+public:
+  deduplicate_operator() = default;
+
+  deduplicate_operator(configuration cfg) : cfg_(std::move(cfg)) {
+  }
+
+  auto operator()(generator<table_slice> input) const
+    -> generator<table_slice> {
+    index_cache cached_indices{};
+    match_store matches{};
+    int64_t row_number{0};
+    int64_t last_cleanup_row{0};
+    auto last_cleanup_time = steady_clock::now();
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto [projected_type, projected_batch] = project(cached_indices, slice);
+      TENZIR_ASSERT(projected_batch);
+      auto elements = projected_batch->ToStructArray().ValueOrDie();
+      TENZIR_ASSERT(elements);
+      for (auto&& new_slice :
+           deduplicate(matches, row_number, slice, projected_type, *elements)) {
+        co_yield std::move(new_slice);
+      }
```

Comment:
Since this will yield a lot of empty slices whenever we remove a duplicate, should we only yield non-empty slices here?